pkg/oc/admin/prune/imageprune/prune.go

```diff
@@ -1011,170 +1132,385 @@ func tagEventIsPrunable(
-	// we still have references to images.
-	if len(p.algorithm.namespace) > 0 || err != nil {
-		return err
-	}
+	// Sorting images from the largest (by number of layers) to the smallest is supposed to distribute the
+	// blob deletion workload equally across whole queue.
+	// If processed randomly, most probably, job processed in the beginning wouldn't delete any blobs (due to
+	// too many remaining referers) contrary to the jobs processed at the end.
+	// The assumption is based on another assumption that images with many layers have a low probability of
+	// sharing their components with other images.
+	sort.Sort(byLayerCountAndAge(p.queue))
 
-	var errs []error
+	var (
+		jobChan    = make(chan *Job)
+		resultChan = make(chan JobResult)
+	)
 
-	if p.algorithm.pruneRegistry {
-		prunableComponents := getPrunableComponents(p.g, prunableImageIDs)
-		errs = append(errs, pruneImageComponents(p.g, p.registryClient, p.registryURL, prunableComponents, layerLinkPruner)...)
-		errs = append(errs, pruneBlobs(p.g, p.registryClient, p.registryURL, prunableComponents, blobPruner)...)
-		errs = append(errs, pruneManifests(p.g, p.registryClient, p.registryURL, prunableImageNodes, manifestPruner)...)
-
-		if len(errs) > 0 {
-			// If we had any errors deleting layers, blobs, or manifest data from the registry,
-			// stop here and don't delete any images. This way, you can rerun prune and retry
-			// things that failed.
-			return kerrors.NewAggregate(errs)
+	defer close(jobChan)
+
+	for i := 0; i < p.numWorkers; i++ {
+		worker, err := NewWorker(
+			p.algorithm,
+			p.registryClientFactory,
+			p.registryURL,
+			imagePrunerFactory,
+			streamPruner,
+			layerLinkPruner,
+			blobPruner,
+			manifestPruner,
+		)
+		if err != nil {
+			failures = append(failures, Failure{
+				Err: fmt.Errorf("failed to initialize worker: %v", err),
+			})
+			return
 		}
+		go worker.Run(jobChan, resultChan)
 	}
 
-	errs = pruneImages(p.g, prunableImageNodes, imagePruner)
-	return kerrors.NewAggregate(errs)
+	ds, fs := p.runLoop(jobChan, resultChan)
+	deletions = append(deletions, ds...)
+	failures = append(failures, fs...)
+
+	return
 }
 
-// imageComponentIsPrunable returns true if the image component is not referenced by any images.
-func imageComponentIsPrunable(g genericgraph.Graph, cn *imagegraph.ImageComponentNode) bool {
-	for _, predecessor := range g.To(cn) {
-		glog.V(4).Infof("Examining predecessor %#v of image config %v", predecessor, cn)
-		if g.Kind(predecessor) == imagegraph.ImageNodeKind {
-			glog.V(4).Infof("Config %v has an image predecessor", cn)
-			return false
+// runLoop processes the queue of prunable images until empty. It makes the workers busy and updates the graph
+// with each change.
+func (p *pruner) runLoop(
+	jobChan chan<- *Job,
+	resultChan <-chan JobResult,
+) (deletions []Deletion, failures []Failure) {
+	imgUpdateChan := p.imageWatcher.ResultChan()
+	isUpdateChan := p.imageStreamWatcher.ResultChan()
```

Comment:
what's the purpose of adding these watches?  Trying to handle the case that a reference to a layer is added to an image or imagestream while we're in the middle of pruning?  It seems unrelated to the fundamental purpose of the PR (to parallelize the image pruning operations)

Reply:
> what's the purpose of adding these watches? Trying to handle the case that a reference to a layer is added to an image or imagestream while we're in the middle of pruning?

exactly

> It seems unrelated to the fundamental purpose of the PR

Yes, it's unrelated, but the way the code is structured, it's now easy to add that safety mechanism. From the past we know that pruning can take hours to complete and that's way too big window for changes and inconsistencies to happen.
For the sake of simplicity (or rather lesser complexity), I can extract that and move to a follow-up if desired.

Reply:
it's fine, just wanted to understand why it was being done here, thanks.